Design.MD

```diff
@@ -0,0 +1,85 @@
+## Use cases
+### Async programming
+### Generators
+## Implementation
+### Api
+#### Public
+
+```scala
+trait Continuation[A]
+  val context: ExecutionContext
+  def resume(value: A): Unit
+  def resumeWithException(error: Throwable): Unit
+```
+
+#### Internal
+
+Some notes studying the Kotlin implementation below.
+
+##### Continuation API
+
+Continuation has two methods, `resume` and `resumeWithException`.
+The first one is to return a _suspended value_,
+and the latter is to _bridge_ an async exception to caller thread.
+Kotlin uses `Result<A>` to merge both methods into 1, but it causes additional overhead and complexity.
+
+```scala
+trait Continuation[A]
+  def resume(value: A): Unit
+  def resumeWithException(error: Throwable): Unit
+```
+
+##### Interception
+
+Kotlin uses `ContinuationIntercepted`, and `Continuation#intercepted` for dispatching.
+The reason for this is that `Continuation` exists in Kotlin Std, but the concurrency implementation lives inside `KotlinX Coroutines`.
+The Kotlin Standard Library has _no concept_ of concurrency **at all**, so no abstraction such as `ExecutionContext` is available.
+
+Using `ExecutionContext` for the Scala implementation makes more sense, so we can replace `CoroutineContext` with `ExecutionContext`.
+`CoroutineContext` is further utilised to implement things such as `CoroutineContextLocal` alla `ThreadContextLocal`,
+and storing other information inside such as `CoroutineName` for debugging purposes so we can still maintain some `State` in a `CoroutineContext` like type.
+We will however not use it for _dispatching_ or scheduling for that we prefer to use `ExecutionContext` in Scala.
+
+```scala
+trait Continuation[A]
+  val context: ExecutionContext
+```
+
+##### Coroutine Builders
+
+Kotlin splits up the builders in `suspendCoroutine`, and `suspendCoroutineCancellable`.
+**Again** because Kotlin Std has no notion of concurrency, and thus no-notion of cancellation.
+We should probably not make this distinction/split in Scala, and prefer to have `suspendCoroutineCancellable` only.
```

Comment:
A side note here is that we need to consider our impl of `Structured` and see if we will support other strategies besides shutdown on failure like [ShutdownOnSuccess](https://download.java.net/java/early_access/loom/docs/api/jdk.incubator.concurrent/jdk/incubator/concurrent/StructuredTaskScope.ShutdownOnSuccess.html).
Also, how the cancellation wiring in these builders interacts with structured concurrency. Since we no longer depend on LOOM in the new model, we have to reimplement `Structured` or the equivalent of `coroutineScope`.

Reply:
This function is like [IO.cancellable](https://typelevel.org/cats-effect/docs/2.x/datatypes/io#cancelable-processes) in the sense that it wraps a single function with cancellation mechanics.

I.e.

```kotlin
suspend fun sdkWrapper(): Int =
  suspendCoroutineCancellable { cont ->
     val callback = Callback { value, exception ->
       if(exception != null) cont.resumeWithException(exception)
       else cont.resume(value)
     }
     sdk.register(callback)
     CancelToken { sdk.unregister(callback) }
  }
```

So this is not related to structured concurrency perse, but we probably do want to have some `structured`/`coroutineScope` alternative with more strategies than KotlinX supports, and probably a saner default / syntax than Loom 😁